### PR TITLE
Completed Add Transactions and few other UI tweaks

### PR DESCRIPTION
**Pull request by akashsarkar188**
Completed Transactions functionality
Structured Dashboard.java
added version to sharedPref
Verify Sheet everytime app opens

### CODE REVIEW

File: app/src/main/java/com/aspirebudgetingmobile/aspirebudgeting/fragments/Transaction.java [hunk 1]
```diff
@@ -1,29 +1,254 @@
...
+import com.google.android.material.card.MaterialCardView;
+
+import java.text.SimpleDateFormat;
+import java.util.ArrayList;
+import java.util.Calendar;
+import java.util.List;
+import java.util.Locale;
+import java.util.Objects;
 
 public class Transaction extends Fragment {
```
Comment (mohitathwani) [inline app/src/main/java/com/aspirebudgetingmobile/aspirebudgeting/fragments/Transaction.java:39]: This class represents a view, please rename it to AddTransactionView or AddTransactionFragment. The name `Transaction` can be misunderstood as a model object.

File: app/src/main/java/com/aspirebudgetingmobile/aspirebudgeting/interfaces/TransactionCallBack.java [hunk 2]
```diff
@@ -0,0 +1,6 @@
+package com.aspirebudgetingmobile.aspirebudgeting.interfaces;
+
+public interface TransactionCallBack {
```
Comment (mohitathwani) [inline app/src/main/java/com/aspirebudgetingmobile/aspirebudgeting/interfaces/TransactionCallBack.java:3]: `TransactionCallBack` is too vague. Rename class to `AddTransactionCallBack` so it is easy to understand when this callback is triggered.

File: app/src/main/java/com/aspirebudgetingmobile/aspirebudgeting/utils/SheetsManager.java [hunk 3]
```diff
@@ -128,18 +149,151 @@ public boolean verifySheet(Context context, String sheetID) {
...
         }
     }
 
     private boolean isSheetVersionSupported(String version) {
 
         List<String> supportedVersion = new ArrayList<>();
-        supportedVersion.add("3.1.0");
-        supportedVersion.add("3.0");
-        supportedVersion.add("2.8");
+        supportedVersion.add(threeOne);
```
Comment (mohitathwani) [inline app/src/main/java/com/aspirebudgetingmobile/aspirebudgeting/utils/SheetsManager.java:160]: add 3.2 to the list.

File: app/src/main/java/com/aspirebudgetingmobile/aspirebudgeting/utils/SheetsManager.java [hunk 4]
```diff
@@ -128,18 +149,151 @@ public boolean verifySheet(Context context, String sheetID) {
...
+                range = "BackendData!M2:M";
+                break;
+        }
+
+        List<List<Object>> data = fetchData(context, sheetID, range);
+        transactionAccounts = convertToOneDimension(data);
+    }
+
+    private List<String> convertToOneDimension(List<List<Object>> data) {
+        List<String> result = new ArrayList<>();
```
Comment (mohitathwani) [inline app/src/main/java/com/aspirebudgetingmobile/aspirebudgeting/utils/SheetsManager.java:212]: Aren't there `filter` and `map` operations in Android/Java?

File: app/src/main/java/com/aspirebudgetingmobile/aspirebudgeting/utils/SheetsManager.java [hunk 5]
```diff
@@ -128,18 +149,151 @@ public boolean verifySheet(Context context, String sheetID) {
...
+            valuesToInsert.add("");
+            valuesToInsert.add(amount);
+        } else {
+            valuesToInsert.add(amount);
+            valuesToInsert.add("");
+        }
+
+        valuesToInsert.add(category);
+        valuesToInsert.add(account);
+        valuesToInsert.add(String.format("(%s) - Added from Aspire Android app", memo));
```
Comment (mohitathwani) [inline app/src/main/java/com/aspirebudgetingmobile/aspirebudgeting/utils/SheetsManager.java:262]: Remove `(` and `)` around `%s`

**Comment by mohitathwani** [comment]
@akashsarkar188 There's actually no reason to verify the sheet every time the app opens. The reason why I do this in iOS is because the current architecture of iOS is somewhat incorrect and is something I am going to fix for version 2.00 of the app.

File: app/src/main/java/com/aspirebudgetingmobile/aspirebudgeting/fragments/Transaction.java (hunk 1 above)
Comment (akashsarkar188) [reply app/src/main/java/com/aspirebudgetingmobile/aspirebudgeting/fragments/Transaction.java:39]: done

File: app/src/main/java/com/aspirebudgetingmobile/aspirebudgeting/interfaces/TransactionCallBack.java (hunk 2 above)
Comment (akashsarkar188) [reply app/src/main/java/com/aspirebudgetingmobile/aspirebudgeting/interfaces/TransactionCallBack.java:3]: done

File: app/src/main/java/com/aspirebudgetingmobile/aspirebudgeting/utils/SheetsManager.java (hunk 3 above)
Comment (akashsarkar188) [reply app/src/main/java/com/aspirebudgetingmobile/aspirebudgeting/utils/SheetsManager.java:160]: done

File: app/src/main/java/com/aspirebudgetingmobile/aspirebudgeting/utils/SheetsManager.java (hunk 4 above)
Comment (akashsarkar188) [reply app/src/main/java/com/aspirebudgetingmobile/aspirebudgeting/utils/SheetsManager.java:212]: I tried finding but couldnt find any, or I am not aware of the method so I used this one.

File: app/src/main/java/com/aspirebudgetingmobile/aspirebudgeting/utils/SheetsManager.java (hunk 5 above)
Comment (akashsarkar188) [reply app/src/main/java/com/aspirebudgetingmobile/aspirebudgeting/utils/SheetsManager.java:262]: done

**Comment by akashsarkar188** [comment]
> 
> 
> @akashsarkar188 There's actually no reason to verify the sheet every time the app opens. The reason why I do this in iOS is because the current architecture of iOS is somewhat incorrect and is something I am going to fix for version 2.00 of the app.

Currently, I implemented it because if I dont then people will have to clear data of the app first else add transactions wont work as I am saving the version in sharedPref, I can remove that in next version.

File: app/src/main/java/com/aspirebudgetingmobile/aspirebudgeting/utils/SheetsManager.java (hunk 3 above)
Comment (akashsarkar188) [reply app/src/main/java/com/aspirebudgetingmobile/aspirebudgeting/utils/SheetsManager.java:176]: yes

**Pull request merged**